File: Univ-5.1/FSM/lab1/task1.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <map>
#include <algorithm>
#include <unordered_set>
#include <set>
#include <algorithm>
#include <iterator>

typedef std::tuple<int, char> tt;

class FiniteStateMachine {
public:
    FiniteStateMachine(
            int _states_count,
            std::vector<char> _alphabet,
            std::map<tt, int> _transition,
            int _start,
            std::vector<int> _final
    ) {

        for (int i = _states_count; i > 0; --i) {
            states.push_back(i - 1);
        }
        transition = std::move(_transition);
        alphabet = std::move(_alphabet);
        start_state = _start;
        std::unordered_set<int> final_states1(_final.begin(), _final.end());
        final_states = final_states1;
    }

    int start_state;
    std::unordered_set<int> final_states;

    bool contains_word(std::string word) {
        int cur_state = start_state;
        while (!word.empty()) {
            cur_state = transit(cur_state, word[0]);
            if (cur_state == -1) return false;
            word = word.substr(1);
        }
        if (final_states.count(cur_state))
            return true;
        return false;
    }

    bool contains_fsm(FiniteStateMachine fsm) {
        std::set<std::pair<int, int>> W;
        std::set<std::pair<int, int>> C;
        W.insert(std::make_pair(start_state, fsm.start_state));
        while (!W.empty()) {
            auto a_b = *W.begin();
            W.erase(W.begin());
            C.insert(a_b);
            if (final_states.count(a_b.first) && final_states.count(a_b.second))
                return false;
            for (auto const &x: alphabet) {
                int _a = transit(a_b.first, x);
                int _b = fsm.transit(a_b.second, x);
                if (_a == -1 && _b != -1) return false;
                if (_a == -1 || _b == -1) continue;
                if (C.find(std::make_pair(_a, _b)) == C.end()) W.insert(std::make_pair(_a, _b));
            }
        }
        return true;
    }


    int transit(int state, char letter) {
        if (transition.count(std::make_tuple(state, letter)))
            return transition[std::make_tuple(state, letter)];
        return -1;
    }

private:
    std::vector<int> states;
    std::vector<char> alphabet;
    std::map<tt, int> transition;

};

int main() {
    auto t1 = std::chrono::high_resolution_clock::now();
    std::vector<bool> ress;
    auto *fsm0 = new FiniteStateMachine(
            4,
            {'a', 'b', 'c'},
            {
                    {std::make_tuple(0, 'a'), 1}
            },
            0,
            {1, 2, 3}
    );

    auto *fsm1 = new FiniteStateMachine(
            4,
            {'a', 'b', 'c'},
            {
                    {std::make_tuple(0, 'a'), 1},
                    {std::make_tuple(1, 'b'), 2},
                    {std::make_tuple(2, 'a'), 2},
                    {std::make_tuple(2, 'c'), 3}
            },
            0,
            {1, 2, 3}
    );

    auto *fsm2 = new FiniteStateMachine(
            4,
            {'a', 'b', 'c', 'd'},
            {
                    {std::make_tuple(0, 'a'), 1},
                    {std::make_tuple(1, 'b'), 2},
                    {std::make_tuple(1, 'c'), 1},
                    {std::make_tuple(1, 'd'), 1},
                    {std::make_tuple(2, 'a'), 2},
                    {std::make_tuple(2, 'c'), 3}
            },
            0,
            {1, 2, 3}
    );

    auto *fsm3 = new FiniteStateMachine(
            5,
            {'a', 'b', 'c', 'd'},
            {
                    {std::make_tuple(0, 'a'), 1},
                    {std::make_tuple(1, 'b'), 2},
                    {std::make_tuple(1, 'c'), 1},
                    {std::make_tuple(1, 'd'), 1},
                    {std::make_tuple(2, 'a'), 2},
                    {std::make_tuple(2, 'c'), 3},
                    {std::make_tuple(2, 'd'), 4},
                    {std::make_tuple(3, 'a'), 0}
            },
            0,
            {1, 2, 3, 4}
    );

    // std::cout << "Check for \"abacd\"\n";
// abaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaac
    for (int i=0; i <1000000; i++ ){
        ress.push_back(fsm3->contains_word("abaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaac"));
        // ress.push_back(fsm1->contains_fsm(*fsm0));
        // auto res = fsm2->contains_fsm(*fsm0);
        // auto res = fsm1->contains_word("abc");

        // std::cout << fsm1->contains_word("abc") << '\n';
        // std::cout << fsm1->contains_fsm(*fsm0) << '\n';
    }

    // std::cout << fsm1->contains_fsm(*fsm0) << '\n';
    // std::cout << fsm3->contains_word("accdbc") << '\n';
    auto t2 = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>( t2 - t1 ).count();
    std::cout << "Executed 100000 times in " << duration << " μs\n";
    std::cout << "Average time " << duration/1000000 << " μs\n";
    std::cout << "Result: " << ress[0];
    return 0;
}

File: Univ-5.1/FSM/lab1/task1.py
import datetime

class FiniteStateMachine:

    def __init__(self, states_count, alphabet, transition, start, final):
        self.states = list(range(states_count))
        self.alphabet = alphabet
        self.transition = {(i, t): v for i, t, v in transition}
        self.start_state = start
        self.final_states = final
        self.cur_state = None

    def transit(self, state, letter):
        try:
            return self.transition[(state, letter)]
        except KeyError:
            return

    def __contains__(self, item):
        if isinstance(item, str):
            return self._contains_word(item)
        if isinstance(item, type(self)):
            return self._contains_fsm(item)

    def _contains_word(self, word):
        cur_state = self.start_state
        while word:
            cur_state = self.transit(cur_state, word[0])
            if cur_state is None: return False
            word = word[1:]
        if cur_state in self.final_states: return True
        return False

    def _contains_fsm(self, fsm):
        C = []
        W = [(self.start_state, fsm.start_state)]
        while W:
            a, b = W.pop(0)
            C.append((a, b))
            if a in self.final_states and b not in fsm.final_states:
                return False
            for x in self.alphabet:
                _a = self.transit(a, x)
                _b = fsm.transit(b, x)
                if _a is None and _b is not None:
                    return False
                if _a is None or _b is None:
                    continue
                if (_a, _b) not in C:
                    W.append((_a, _b))
        return True


thelist = []

params0 = {
    'states_count': 4,
    'alphabet': ['a', 'b', 'c'],
    'transition': [
        [0, 'a', 1],
    ],
    'start': 0,
    'final': [1, 2, 3]
}
fsm0 = FiniteStateMachine(**params0)

params1 = {
    'states_count': 4,
    'alphabet': ['a', 'b', 'c'],
    'transition': [
        [0, 'a', 1],
        [1, 'b', 2],
        [2, 'a', 2],
        [2, 'c', 3]
    ],
    'start': 0,
    'final': [1, 2, 3]
}
fsm1 = FiniteStateMachine(**params1)

params2 = {
    'states_count': 4,
    'alphabet': ['a', 'b', 'c', 'd'],
    'transition': [
        [0, 'a', 1],
        [1, 'c', 1],
        [1, 'd', 1],
        [1, 'b', 2],
        [2, 'a', 2],
        [2, 'c', 3]
    ],
    'start': 0,
    'final': [1, 2, 3]
}
fsm2 = FiniteStateMachine(**params2)

params3 = {
    'states_count': 5,
    'alphabet': ['a', 'b', 'c', 'd'],
    'transition': [
        [0, 'a', 1],
        [1, 'c', 1],
        [1, 'd', 1],
        [1, 'b', 2],
        [2, 'a', 2],
        [2, 'c', 3],
        [2, 'd', 4],
        [3, 'a', 0]
    ],
    'start': 0,
    'final': [1, 2, 3, 4]
}
fsm3 = FiniteStateMachine(**params3)

# print("Check for \"abc\"")
start_time = datetime.datetime.now()

for i in range(1000000):
    # thelist.append(('abc' in fsm1, fsm0 in fsm1))  
    # thelist.append(fsm0 in fsm1)  
    thelist.append('abaaaaaaaaac' in fsm3)   

duration = (datetime.datetime.now() - start_time).total_seconds() * 1000000
print("Executed {} times in {} μs".format(1000000, duration))
print("Average time {} μs".format(duration/1000000))
print("Result: {}".format(thelist[0]))

# print('abc' in fsm1)  # return True
# print(fsm0 in fsm1)  # return True